jsonComposer.go

```diff
@@ -0,0 +1,218 @@
+	}
+
+	if dst == nil {
+		dst = make([]byte, len(src))
+	}
+
+	composer := &jsonObjComposer{
+		body:                dst,
+		pos:                 0,
+		atLeastOneFieldLeft: false,
+	}
+
+	tokenizer := &jsonTokenizer{}
+	tokenizer.Reset(src)
+
+	var tknType tokenType
+	var tknLen, dstLen int
+	var tkn []byte
+	var err error
+	var atleastOneFieldLeft, removedLastItem bool
+	var depth int
+	for tknType != tknEnd {
+		tknType, tkn, tknLen, err = tokenizer.Step()
+		if err != nil {
+			return 0, 0, false, fmt.Errorf("error stepping to next token, src=%s, pos=%v, err=%v", src, tokenizer.Position(), err)
+		}
+
+		switch tknType {
+		case tknString:
+			// depth 1 strings are JSON object keys, need to check if it is the key to remove
+			if depth != 1 {
+				composer.Write(tkn)
+				continue
+			}
+		case tknObjectStart:
+			fallthrough
+		case tknArrayStart:
+			depth++
+			composer.Write(tkn)
+			continue
+		case tknObjectEnd:
+			fallthrough
+		case tknArrayEnd:
+			depth--
+			if depth < 0 {
+				return 0, 0, false, fmt.Errorf("invalid JSON object")
+			}
+			composer.Write(tkn)
+			continue
+		case tknEnd:
+			continue
+		default:
+			composer.Write(tkn)
+			continue
+		}
+
+		// strip off the quotes from the string
+		key := tkn[1 : tknLen-1]
+
+		matched := false
+		for _, keyToRemove := range remove {
+			if BytesEqualsString(key, keyToRemove) {
+				matched = true
+
+				// ":"
+				tknType, _, _, err = tokenizer.Step()
+				if err != nil || tknType != tknObjectKeyDelim {
+					return 0, 0, false, fmt.Errorf("error stepping to next token, expecting :, src=%s, pos=%v, err=%v", src, tokenizer.Position(), err)
+				}
+
+				// parse the corresponding value
+				valStart := tokenizer.Position()
+				valEnd := valStart
+				valueDepth := 0
+				done := false
+
+				for !done {
+					tknType, _, _, err = tokenizer.Step()
+					if err != nil {
+						return 0, 0, false, fmt.Errorf("error stepping to next token, expecting JSON value, src=%s, pos=%v, err=%v", src, tokenizer.Position(), err)
+					}
+
+					valEnd = tokenizer.Position()
+
+					switch tknType {
+					case tknObjectStart:
+						fallthrough
+					case tknArrayStart:
+						valueDepth++
+					case tknObjectEnd:
+						fallthrough
+					case tknArrayEnd:
+						valueDepth--
+						if valueDepth == 0 {
+							done = true
+						} else if valueDepth < 0 {
+							return 0, 0, false, fmt.Errorf("invalid JSON object, src=%s, pos=%v", src, tokenizer.Position())
+						}
+					case tknString:
+						fallthrough
+					case tknEscString:
+						fallthrough
+					case tknInteger:
+						fallthrough
+					case tknNumber:
+						fallthrough
+					case tknNull:
+						fallthrough
+					case tknTrue:
+						fallthrough
+					case tknFalse:
+						if valueDepth == 0 {
+							done = true
+						}
+					case tknEnd:
+						return 0, 0, false, fmt.Errorf("unexpected EOF, src=%s, pos=%v", src, tokenizer.Position())
+					default:
+						// tknListDelim, tknObjectKeyDelim, tknUnknown
```

Comment:
What is this line supposed to mean?

Reply:
The default case is hit when it is either of tknListDelim, tknObjectKeyDelim or tknUnknown - no action need to be taken.